databases/core.py

```diff
@@ -113,7 +112,10 @@ async def disconnect(self) -> None:
             self._global_transaction = None
             self._global_connection = None
         else:
-            self._connection_context = ContextVar("connection_context")
+            current_task = asyncio.current_task()
+            assert current_task is not None, "No currently running task"
+            if current_task in self._connections:
+                del self._connections[current_task]
```

Comment:
Maybe I'm missing something, but this appears to the only place where items are removed from the _connections dict. So if I'm running a web service that calls connect() when the service starts up and disconnect() when it shuts down, but in the interim I'm serving each request in a new task, I'd have an entry in the _connections dictionary for every request that used the database until my service shuts down. Also, because the current task is used as a key in the dict any request that used the database won't have it's task state garbage collected, as the dict is holding a reference to it.

Reply:
Oh I see that now too - thank you! New tasks do get allocated a new connection, and they aren't necessarily going to each call disconnect before they complete.

- Main task calls `db.connect`, a new `Connection` is created for it
- Main task creates another task 
- Child task has a new `Connection` allocated because it uses `db.execute`, or alike
- Child task does not call `db.disconnect` before it is completed
- Child task completes
- `db._connections` still has the child task as a key referencing the child's connection.

Reply:
👍 thanks — this should be resolved by not caching per task and instead leveraging context variables. Otherwise you're right that a weakkey dict makes sense.